### PR TITLE
Filter documents with iterators

### LINKED ISSUE #1733: Support `getMore` command
### What should be done?

https://www.mongodb.com/docs/manual/reference/command/getMore/

- [x] Do not consume all documents in `find`: #2933
- [x] Do not consume all documents in `aggregate`
- [ ] Refactor tests (https://github.com/FerretDB/FerretDB/pull/2929/files#r1246365875)

Test script: #2761

### PR DESCRIPTION
**Pull request by AlekSi**
# Description

Closes #1733.

## Readiness checklist

* [x] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [x] I added/updated comments and checked rendering.
* [x] I made spot refactorings.
* [ ] I updated user documentation.
* [x] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/clientconn/conninfo/conn_info.go [hunk 1]
```diff
@@ -116,7 +114,7 @@ func (connInfo *ConnInfo) Cursor(id int64) *Cursor {
 }
 
 // StoreCursor stores cursor and return its ID.
-func (connInfo *ConnInfo) StoreCursor(iter iterator.Interface[int, *types.Document], filter *types.Document) int64 {
+func (connInfo *ConnInfo) StoreCursor(iter types.DocumentsIterator) int64 {
```
Comment (rumyantseva) [inline internal/clientconn/conninfo/conn_info.go:117]: @AlekSi do you understand why some lines are marked as not covered by tests? E.g. this function - the signature has changes, so it has changed in the callers too, but where are those callers?

File: internal/handlers/common/filter_iterator.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	iter   types.DocumentsIterator
+	filter *types.Document
+}
+
+// Next implements iterator.Interface. See FilterIterator for details.
+func (iter *filterIterator) Next() (int, *types.Document, error) {
+	for {
+		i, doc, err := iter.iter.Next()
+		if err != nil {
+			return 0, nil, lazyerrors.Error(err)
```
Comment (rumyantseva) [inline internal/handlers/common/filter_iterator.go:48]: So, even for the case when iterator is done, we return a lazy error? Should we follow this approach everywhere else as well?

File: internal/types/documents_iterator.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import "github.com/FerretDB/FerretDB/internal/util/iterator"
+
+// DocumentsIterator represents an iterator over documents.
```
Comment (rumyantseva) [inline internal/types/documents_iterator.go:19]: Should we specify what `key` means in this case?

**Review by rumyantseva — COMMENTED** [review]
I understand the idea behind DocumentsIterator and FilterIterator, and I'm ok with it. Overall, with all those iterators, the code looks more and more complicated, but it looks like those things are needed for us.

Asked a few questions.

File: internal/clientconn/conninfo/conn_info.go (hunk 1 above)
Comment (AlekSi) [reply internal/clientconn/conninfo/conn_info.go:117]: There are none. This function will be used in future PRs.

File: internal/handlers/common/filter_iterator.go (hunk 2 above)
Comment (AlekSi) [reply internal/handlers/common/filter_iterator.go:48]: > // Returned error could be (possibly wrapped) ErrIteratorDone or some fatal error
> // like (possibly wrapped) context.Canceled.

So the caller should be able to handle both. And there it allows us to simplify code

File: internal/types/documents_iterator.go (hunk 3 above)
Comment (AlekSi) [reply internal/types/documents_iterator.go:19]: That's undefined in the type itself, and documented by users of it

**Review by rumyantseva — APPROVED** [review]

**Review by w84thesun — APPROVED** [review]

**Pull request merged**